File: examples/visitor-pattern/async-validation.py
#!/usr/bin/env python3
"""
Asynchronous URL Validation Example.

Demonstrates how to validate URLs asynchronously during conversion.
Uses aiohttp to check HTTP status codes of all links and images.
Fails conversion if any URLs are broken (404, 500, etc.).
"""

import asyncio

import aiohttp

from html_to_markdown import convert_with_async_visitor


class AsyncUrlValidator:
    """Validates all URLs asynchronously using HTTP HEAD requests."""

    def __init__(self, timeout: int = 5) -> None:
        self.timeout = timeout
        self.session: aiohttp.ClientSession | None = None
        self.validated_urls = []

    async def visit_link(self, ctx, href: str, text: str, title: str | None):
        """Validate link URLs."""
        # Skip anchor links
        if href.startswith("#"):
            return {"type": "continue"}

        # Skip mailto and tel links
        if href.startswith(("mailto:", "tel:")):
            return {"type": "continue"}

        # Validate HTTP(S) URLs
        if href.startswith(("http://", "https://")):
            is_valid, status = await self._validate_url(href)
            self.validated_urls.append((href, status))

            if not is_valid:
                return {"type": "error", "message": f"Broken link ({status}): {href}"}

        return {"type": "continue"}

    async def visit_image(self, ctx, src: str, alt: str | None, title: str | None):
        """Validate image URLs."""
        # Skip data URIs and relative paths
        if not src.startswith(("http://", "https://")):
            return {"type": "continue"}

        is_valid, status = await self._validate_url(src)
        self.validated_urls.append((src, status))

        if not is_valid:
            return {"type": "error", "message": f"Broken image ({status}): {src}"}

        return {"type": "continue"}

    async def _validate_url(self, url: str) -> tuple[bool, int | str]:
        """Validate a URL using HTTP HEAD request."""
        if self.session is None:
            self.session = aiohttp.ClientSession()

        try:
            async with self.session.head(url, timeout=self.timeout, allow_redirects=True) as response:
                # Consider 2xx and 3xx as valid
                is_valid = response.status < 400
                return (is_valid, response.status)
        except asyncio.TimeoutError:
            return (False, "timeout")
        except aiohttp.ClientError as e:
            return (False, f"error: {type(e).__name__}")
        except Exception as e:
            return (False, f"error: {e!s}")

    async def close(self) -> None:
        """Close the HTTP session."""
        if self.session:
            await self.session.close()


async def test_valid_urls() -> bool | None:
    """Test with valid URLs (should pass)."""
    html = """
    <h1>Article with Valid Links</h1>
    <p>Check out <a href="https://www.example.com">Example</a>.</p>
    <img src="https://via.placeholder.com/150" alt="Placeholder image">
    <p>Internal anchor: <a href="#section1">Section 1</a></p>
    """

    visitor = AsyncUrlValidator(timeout=10)
    try:
        await convert_with_async_visitor(html, visitor=visitor)
        for _url, _status in visitor.validated_urls:
            pass
        return True
    except Exception:
        return False
    finally:
        await visitor.close()


async def test_broken_link() -> bool | None:
    """Test with broken URL (should fail)."""
    html = """
    <h1>Article with Broken Link</h1>
    <p>This link is broken: <a href="https://httpstat.us/404">404 page</a>.</p>
    """

    visitor = AsyncUrlValidator(timeout=10)
    try:
        await convert_with_async_visitor(html, visitor=visitor)
        return False
    except Exception:
        for _url, _status in visitor.validated_urls:
            pass
        return True
    finally:
        await visitor.close()


async def main() -> None:
    tests = [
        ("Valid URLs", test_valid_urls),
        ("Broken Link (404)", test_broken_link),
    ]

    results = []
    for name, test_func in tests:
        passed = await test_func()
        results.append((name, passed))

    for name, passed in results:
        pass


if __name__ == "__main__":
    asyncio.run(main())

File: examples/visitor-pattern/async-validation.ts
#!/usr/bin/env tsx
/**
 * Asynchronous URL Validation Example (TypeScript)
 *
 * Demonstrates how to validate URLs asynchronously during conversion.
 * Uses fetch to check HTTP status codes of all links and images.
 * Fails conversion if any URLs are broken (404, 500, etc.).
 */

import { convertWithAsyncVisitor, type AsyncVisitor, type NodeContext, type VisitResult } from 'html-to-markdown';

class AsyncUrlValidator implements AsyncVisitor {
  private timeout: number;
  public validatedUrls: Array<[string, number | string]> = [];

  constructor(timeout: number = 5000) {
    this.timeout = timeout;
  }

  async visitLink(ctx: NodeContext, href: string, text: string, title?: string): Promise<VisitResult> {
    // Skip anchor links
    if (href.startsWith('#')) {
      return { type: 'continue' };
    }

    // Skip mailto and tel links
    if (href.startsWith('mailto:') || href.startsWith('tel:')) {
      return { type: 'continue' };
    }

    // Validate HTTP(S) URLs
    if (href.startsWith('http://') || href.startsWith('https://')) {
      const [isValid, status] = await this.validateUrl(href);
      this.validatedUrls.push([href, status]);

      if (!isValid) {
        return {
          type: 'error',
          message: `Broken link (${status}): ${href}`,
        };
      }
    }

    return { type: 'continue' };
  }

  async visitImage(ctx: NodeContext, src: string, alt?: string, title?: string): Promise<VisitResult> {
    // Skip data URIs and relative paths
    if (!src.startsWith('http://') && !src.startsWith('https://')) {
      return { type: 'continue' };
    }

    const [isValid, status] = await this.validateUrl(src);
    this.validatedUrls.push([src, status]);

    if (!isValid) {
      return {
        type: 'error',
        message: `Broken image (${status}): ${src}`,
      };
    }

    return { type: 'continue' };
  }

  private async validateUrl(url: string): Promise<[boolean, number | string]> {
    try {
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      const response = await fetch(url, {
        method: 'HEAD',
        signal: controller.signal,
        redirect: 'follow',
      });

      clearTimeout(timeoutId);

      // Consider 2xx and 3xx as valid
      const isValid = response.status < 400;
      return [isValid, response.status];
    } catch (error) {
      if (error instanceof Error) {
        if (error.name === 'AbortError') {
          return [false, 'timeout'];
        }
        return [false, `error: ${error.message}`];
      }
      return [false, 'error: unknown'];
    }
  }
}

async function testValidUrls(): Promise<boolean> {
  console.log('Test: Valid URLs');
  console.log('-'.repeat(70));

  const html = `
    <h1>Article with Valid Links</h1>
    <p>Check out <a href="https://www.example.com">Example</a>.</p>
    <img src="https://via.placeholder.com/150" alt="Placeholder image">
    <p>Internal anchor: <a href="#section1">Section 1</a></p>
  `;

  const visitor = new AsyncUrlValidator(10000);
  try {
    const markdown = await convertWithAsyncVisitor(html, { visitor });
    console.log('✓ All URLs are valid');
    console.log();
    console.log('Validated URLs:');
    for (const [url, status] of visitor.validatedUrls) {
      console.log(`  [${status}] ${url}`);
    }
    console.log();
    console.log('Markdown output:');
    console.log('-'.repeat(70));
    console.log(markdown);
    return true;
  } catch (error) {
    console.log(`✗ Unexpected error: ${error}`);
    return false;
  }
}

async function testBrokenLink(): Promise<boolean> {
  console.log('Test: Broken Link (404)');
  console.log('-'.repeat(70));

  const html = `
    <h1>Article with Broken Link</h1>
    <p>This link is broken: <a href="https://httpstat.us/404">404 page</a>.</p>
  `;

  const visitor = new AsyncUrlValidator(10000);
  try {
    const markdown = await convertWithAsyncVisitor(html, { visitor });
    console.log('✗ Should have failed on broken link');
    return false;
  } catch (error) {
    console.log(`✓ Correctly caught broken link: ${error}`);
    console.log();
    console.log('Validated URLs:');
    for (const [url, status] of visitor.validatedUrls) {
      console.log(`  [${status}] ${url}`);
    }
    return true;
  }
}

async function main() {
  console.log('='.repeat(70));
  console.log('Asynchronous URL Validation Example (TypeScript)');
  console.log('='.repeat(70));
  console.log();
  console.log('Note: This example makes real HTTP requests to validate URLs.');
  console.log('Ensure you have an internet connection for the tests to work.');
  console.log();

  const tests: Array<[string, () => Promise<boolean>]> = [
    ['Valid URLs', testValidUrls],
    ['Broken Link (404)', testBrokenLink],
  ];

  const results: Array<[string, boolean]> = [];
  for (const [name, testFunc] of tests) {
    const passed = await testFunc();
    results.push([name, passed]);
    console.log();
  }

  console.log('='.repeat(70));
  console.log('Test Summary:');
  console.log('-'.repeat(70));
  for (const [name, passed] of results) {
    const status = passed ? 'PASS' : 'FAIL';
    console.log(`  ${status}: ${name}`);
  }
  console.log();
}

main();
